File: app/models/annotation.py
"""Pydantic models for annotation records."""

from pydantic import BaseModel


class BBox(BaseModel):
    """Bounding box utility model (x, y, width, height)."""

    x: float
    y: float
    w: float
    h: float


class AnnotationResponse(BaseModel):
    """Single annotation record returned by the API."""

    id: str
    dataset_id: str
    sample_id: str
    category_name: str
    bbox_x: float
    bbox_y: float
    bbox_w: float
    bbox_h: float
    area: float
    is_crowd: bool
    source: str
    confidence: float | None = None


class BatchAnnotationsResponse(BaseModel):
    """Batch annotation response grouping annotations by sample_id."""

    annotations: dict[str, list[AnnotationResponse]]


class AnnotationUpdate(BaseModel):
    """Request body for PUT /annotations/{id} -- update bbox position/size."""

    bbox_x: float
    bbox_y: float
    bbox_w: float
    bbox_h: float


class AnnotationCreate(BaseModel):
    """Request body for POST /annotations -- create a new ground_truth annotation."""

    dataset_id: str
    sample_id: str
    category_name: str
    bbox_x: float
    bbox_y: float
    bbox_w: float
    bbox_h: float


class CategoryUpdateRequest(BaseModel):
    """Request body for PATCH /annotations/{id}/category -- update classification label."""

    category_name: str

File: frontend/src/types/annotation.ts
/**
 * TypeScript types matching backend AnnotationResponse
 * and BatchAnnotationsResponse models from app/models/annotation.py.
 */

export interface Annotation {
  id: string;
  dataset_id: string;
  sample_id: string;
  category_name: string;
  bbox_x: number;
  bbox_y: number;
  bbox_w: number;
  bbox_h: number;
  area: number;
  is_crowd: boolean;
  source: string;
  confidence: number | null;
}

/**
 * Response from the batch annotations endpoint.
 * Maps sample_id to its list of annotations.
 */
export interface BatchAnnotationsResponse {
  annotations: Record<string, Annotation[]>;
}

/** Request body for PUT /annotations/{id} -- update bbox position/size. */
export interface AnnotationUpdate {
  bbox_x: number;
  bbox_y: number;
  bbox_w: number;
  bbox_h: number;
}

/** Request body for POST /annotations -- create a new ground_truth annotation. */
export interface AnnotationCreate {
  dataset_id: string;
  sample_id: string;
  category_name: string;
  bbox_x: number;
  bbox_y: number;
  bbox_w: number;
  bbox_h: number;
}
